Custom Fallback TOML Config

EasterTheBunny opened the pull request:
This commit provides using an existing env var `CL_CHAIN_DEFAULTS` as a path to a custom `fallback.toml`. This allows plugins to define their own set of fallback options apart from the core node which override the default fallback options.

[CCIP-4395](https://smartcontract-it.atlassian.net/browse/CCIP-4395)

[CCIP-4395]: https://smartcontract-it.atlassian.net/browse/CCIP-4395?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

patrickhuie19 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -33,48 +33,83 @@ var (
+	// fallback.toml(defaults dir) <- fallback.toml(env CL_CHAIN_FALLBACK) <- ChainSpecific.toml(env CL_CHAIN_DEFAULTS)
+	//
+	// the custom fallback gets processed and overrides the default fallback
+	if path := env.CustomFallback.Get(); path != "" {
```
can we lock this in with a testscripts test like in https://github.com/smartcontractkit/chainlink/pull/14090/files#diff-ea10429fdb1fbc100e3d8082e39e6d40a12f82b48c1f37ab9158b6c3c0e44d81 ?

EasterTheBunny replied:
Good idea. I'll add that

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -33,48 +32,83 @@ var (
-		id := config.ChainID.String()
+
+		// add ChainID to set of default IDs
+		DefaultIDs = append(DefaultIDs, chainID)
```
I think it is important not to separate these global var assignments (`defaults` and `defaultNames` too) from their declarations.

jmank88 replied:
If you want to use helpers, then how about returning values to set from `init()`?

EasterTheBunny committed:
Custom Fallback TOML Config

This commit provides using an existing env var `CL_CHAIN_DEFAULTS` as a path to a custom `fallback.toml`. This allows
plugins to define their own set of fallback options apart from the core node which override the default fallback
options.

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -98,39 +117,60 @@ func init() {
+func readConfig(path string, reader func(name string) ([]byte, error), fallback bool) (*big.Big, Chain, error) {
+	bts, err := reader(path)
+	if err != nil {
+		return nil, Chain{}, fmt.Errorf("%w: %s", errRead, err.Error())
```
```suggestion
		return nil, Chain{}, fmt.Errorf("%w: %w", errRead, err)
```

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -98,39 +117,60 @@ func init() {
+				continue
+			}
+
+			log.Fatalf("custom defaults override failure (%s): %s", entry.Name(), err.Error())
```
```suggestion
			log.Fatalf("custom defaults override failure (%s): %s", entry.Name(), err)
```

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -98,39 +117,60 @@ func init() {
-		id := config.ChainID.String()
+	// decode from toml to a chain config
+	if err := cconfig.DecodeTOML(bytes.NewReader(bts), &config); err != nil {
+		return nil, Chain{}, fmt.Errorf("%w %s: %s", errDecode, path, err.Error())
```
```suggestion
		return nil, Chain{}, fmt.Errorf("%w %s: %w", errDecode, path, err)
```

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -33,48 +32,68 @@ var (
 
 func init() {
-	// read the defaults first
+	// read all default configs
+	initReadDefaults()
+
+	// check for and apply any overrides
+	initApplyEVMOverrides()
```
I don't think this indirection is adding value. Please consider setting the global vars from this func, one way or another.
https://github.com/smartcontractkit/chainlink/pull/15617/files#r1883872492

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -33,48 +32,68 @@ var (
+}
 
+var (
+	errRead           = errors.New("error reading file")
+	errDecode         = errors.New("error in TOML decoding")
+	errMissingChainID = errors.New("missing ChainID")
+	errNonNilChainID  = errors.New("fallback ChainID must be nil")
+	errFallbackConfig = errors.New("fallback config")
```
Are these all single use? I think the indirection makes it harder to read.

jmank88 replied:
Ah I guess `errFallbackConfig` is actually checked for

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -98,39 +117,60 @@ func init() {
-		if _, ok := customDefaults[id]; ok {
-			log.Fatalf("%q contains duplicate ChainID: %s", path, id)
+	if fallback {
+		if config.ChainID != nil {
+			return nil, Chain{}, fmt.Errorf("%w: found: %s", errNonNilChainID, config.ChainID)
 		}
-		customDefaults[id] = config.Chain
+
+		return nil, config.Chain, errFallbackConfig
```
We should not be using a custom error to communicate back to the caller when there is not an error, and it seems unnecessary in this case since they passed it in to us in the first place.

jmank88 commented on core/chains/evm/config/toml/defaults.go:
```diff
@@ -98,39 +117,60 @@ func init() {
+		chainID, chain, err := readConfig(path, os.ReadFile, entry.Name() == "fallback.toml")
 		if err != nil {
-			log.Fatalf("error opening file (name: %v) in custom defaults override directory: %v", entry.Name(), err)
+			if errors.Is(err, errFallbackConfig) {
```
We already know when it is a fallback config, because we just passed the boolean arg to `readConfig`. Let's use that logic directly instead of hacking errors.

jmank88 commented on pnpm-lock.yaml:
Incidental?

EasterTheBunny replied:
Yup. I don't know why that is there. I can remove it

EasterTheBunny replied:
Now I'm not sure what this is. It must be incidental because I added a new `.txtar` file to handle the new test case. I don't think I can remove it.

jmank88 replied:
What happens if you revert this file? I thought it would only update from a manual command.

EasterTheBunny committed:
collapse helper functions into single helper function and reduce indirection

EasterTheBunny committed:
fix test
